spec/unit/ssl/state_machine_spec.rb

```diff
@@ -56,6 +60,35 @@
     end
   end
 
+  context 'when locking' do
+    it 'locks the file prior to running the state machine' do
+      expect(Puppet::FileSystem).to receive(:exclusive_open).and_call_original.ordered
+      expect(cert_provider).to receive(:load_cacerts).and_return(cacerts).ordered
+      expect(cert_provider).to receive(:load_crls).and_return(crls).ordered
+
+      machine.ensure_ca_certificates
+    end
+
+    it 'raises an exception if it fails to lock' do
+      allow(Puppet::FileSystem.instance_variable_get(:@impl)).to receive(:sleep)
+
+      Puppet::FileSystem.exclusive_open(Puppet[:ssl_lockfile], 0644, 'w', 0) do |fd|
+        expect {
+          machine.ensure_ca_certificates
+        }.to raise_error(Puppet::Error, /Another puppet instance is already running; exiting/)
+      end
+    end
+
+    it 'leaves a lockfile behind' do
+      allow(cert_provider).to receive(:load_cacerts).and_return(cacerts)
+      allow(cert_provider).to receive(:load_crls).and_return(crls)
+
+      expect(File).to_not be_exist(Puppet[:ssl_lockfile])
+      machine.ensure_ca_certificates
+      expect(File).to be_exist(Puppet[:ssl_lockfile])
```

Comment:
Does it ever actually write any contents to the file?  Would something like the pid here be useful for debugging?

Reply:
Yeah, let me add the `pid`

Reply:
Done